src/ir/ranges.cpp

```diff
@@ -0,0 +1,256 @@
+            break;
+        }
+        case 1:    // arg_range_type
+            ++util::get<1>(it_);
+            break;
+        default:
+            HPX_THROW_EXCEPTION(hpx::invalid_status,
+                "phylanx::ir::range_iterator::increment",
+                "range_iterator object holds unsupported data type");
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////////////
+    execution_tree::primitive_argument_type reverse_range_iterator::dereference() const
+    {
+        switch (it_.index())
+        {
+        case 0:    // int_range_type
+            return execution_tree::primitive_argument_type(
+                util::get<0>(it_).first);
+        case 1:    // arg_range_type
+            return *(util::get<1>(it_));
+        }
+
+        HPX_THROW_EXCEPTION(hpx::invalid_status,
+            "phylanx::ir::reverse_range_iterator::dereference",
+            "reverse_range_iterator object holds unsupported data type");
+    }
+
+    bool reverse_range_iterator::equal(reverse_range_iterator const& other) const
+    {
+        // Ensure that have comparable iterators
+        if (it_.index() != other.it_.index())
+        {
+            return false;
+        }
+
+        switch (it_.index())
+        {
+        case 0:    // int_range_type
+            return util::get<0>(it_).first == util::get<0>(other.it_).first;
+        case 1:    // arg_range_type
+            return util::get<1>(it_) == util::get<1>(other.it_);
+        }
+
+        HPX_THROW_EXCEPTION(hpx::invalid_status,
+            "phylanx::ir::reverse_range_iterator::equal",
+            "reverse_range_iterator object holds unsupported data type");
+    }
+
+    void reverse_range_iterator::increment()
+    {
+        switch (it_.index())
+        {
+        case 0:    // int_range_type
+        {
+            int_range_type& p = util::get<0>(it_);
+            p.first -= p.second;
+            break;
+        }
+        case 1:    // arg_range_type
+            ++util::get<1>(it_);
+            break;
+        default:
+            HPX_THROW_EXCEPTION(hpx::invalid_status,
+                "phylanx::ir::reverse_range_iterator::increment",
+                "reverse_range_iterator object holds unsupported data type");
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////////////
+    range_iterator range::begin()
+    {
+        switch (data_.index())
+        {
+        case 0:    // int_range_type
+        {
+            int_range_type& int_range = util::get<0>(data_);
+            std::int64_t start = hpx::util::get<0>(int_range);
+            std::int64_t step = hpx::util::get<2>(int_range);
+            return range_iterator{start, step};
+        }
+        case 1:    // args_type
+            return util::get<1>(data_).begin();
+        case 2:    // arg_pair_type
+            return util::get<2>(data_).first;
+        case 3:
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
```

Comment:
Hmmm. A range should not know whether it is a 'reverse' range or not. Also ranges are always 'forward' ranges. This makes it always be possible to derive a reverse iterator from a forward range.

Reply:
I think I addressed it. We can't do `range x(v.rbegin(), v.rend()` anymore, though.